### PR TITLE
Change the way registration of mutators is done with Replicache.

### PR DESCRIPTION
**Pull request by aboodman**
TypeScript can infer the shape of the Data object if we construct is an object rather than a class, which allows us to avoid typing the signature twice and removes the temptation to abstract out the signature types, removing tons of goop.

This in turn enables removing the mutators.ts file, which was really just a place to hide all the ts goop. There's no need for the distinction between this and the domain objects.

We could actually infer wrap this up more, and general wrapper functions for our mutations that we feed right into `rep.register()`, but I decided it made sense to be explicit here.

It's not that much work, and you should realize when you are changing both these places.

### CODE REVIEW

File: frontend/data.ts [hunk 1]
```diff
@@ -1,101 +1,99 @@
...
-  private rep: Replicache;
-  readonly clientID: string;
+export type Data = ReturnType<typeof createData>;
 
-  constructor(rep: Replicache) {
-    this.rep = rep;
+export function createData(rep: Replicache) {
+  // TODO: Use clientID from Replicache:
+  // https://github.com/rocicorp/replicache-sdk-js/issues/275
+  let clientID = localStorage.clientID;
```
Comment (arv) [inline frontend/data.ts:16]: localStorage is slow. Is this called a lot?

File: frontend/data.ts [hunk 2]
```diff
@@ -1,101 +1,99 @@
...
-    return {
-      getObject: tx.get.bind(tx),
-      putObject: () => {
-        throw new Error("Cannot write inside ReadTransaction");
-      },
-    };
-  }
+function readStorage(tx: ReadTransaction): Storage {
+  return {
+    getObject: tx.get.bind(tx),
```
Comment (arv) [inline frontend/data.ts:88]: bind is slower but 🤷 

I don't think I ever felt the need to use bind once I got used to arrow functions.

File: frontend/data.ts [hunk 3]
```diff
@@ -1,101 +1,99 @@
...
+    putObject: () => {
+      throw new Error("Cannot write inside ReadTransaction");
```
Comment (arv) [inline frontend/data.ts:90]: Why have this method? If you remove it TS will catch errors statically.

File: frontend/data.ts [hunk 4]
```diff
@@ -1,101 +1,99 @@
...
+  };
+}
 
-  private writeStorage(tx: WriteTransaction): Storage {
-    return Object.assign(this.readStorage(tx), {
-      putObject: tx.put.bind(tx),
-    });
-  }
+function writeStorage(tx: WriteTransaction): Storage {
+  return Object.assign(readStorage(tx), {
```
Comment (arv) [inline frontend/data.ts:96]: I find this strange. Maybe just define the functions and do:

```ts
function writeStorage(tx: WriteTransaction): Storage {
  return {getObject, putObject};
}
```

**Review by arv — COMMENTED** [review]
LGTM

File: frontend/data.ts (hunk 1 above)
Comment (aboodman) [reply frontend/data.ts:16]: No, it's just called once on page load.

Also, please expose Replicache's client ID then I wouldn't need this :). Also the user chip wouldn't flicker on startup.

https://github.com/rocicorp/replicache-sdk-js/issues/275

File: frontend/data.ts (hunk 2 above)
Comment (aboodman) [reply frontend/data.ts:88]: I love bind so much, but if it's slower, I'll go with your way.

File: frontend/data.ts (hunk 3 above)
Comment (aboodman) [reply frontend/data.ts:90]: Yeah dunno what I was thinking. Required some refactoring but easy enough. Done.

File: frontend/data.ts (hunk 4 above)
Comment (aboodman) [reply frontend/data.ts:96]: RESPECT MY ART ERIK

**Pull request merged**